In [26]:
import pandas as pd

file_path = r"C:\Users\ADMIN\Documents\mini_project_guvi\project_diamond\regression.csv" 
data = pd.read_csv(file_path)
data.head()

,y,carat,volume,clarity,x,color,price_per_carat,price_inr
0,3.98,0.207014,38.202030,2,3.95,6,27.968908,480.566694
1,3.84,0.190620,34.505856,3,3.89,6,30.374301,480.566694
2,4.07,0.207014,38.076885,5,4.05,6,27.983657,480.820129
3,4.23,0.254642,46.724580,4,4.20,2,22.832547,482.572834
4,4.35,0.270027,51.917250,2,4.34,1,21.542691,482.820226


# Train test split

In [27]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler


X = data.drop("price_inr", axis=1)   # input features
y = data["price_inr"]        # target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Feature Scaling 
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


# Model Building

In [28]:
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.neighbors import KNeighborsRegressor
from xgboost import XGBRegressor


# ML Regression Models
models = {
    "Linear Regression": LinearRegression(),
    "Random Forest": RandomForestRegressor(n_estimators=50, random_state=42),
    "Decision Tree": DecisionTreeRegressor(random_state=42),
    "XGBoost": XGBRegressor(n_estimators=50, random_state=42, verbosity=0),
    "KNN": KNeighborsRegressor()
}


# Evaluate Model

In [29]:
import numpy as np
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

results = {}

for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    
    mae = mean_absolute_error(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)
    rmse = np.sqrt(mse)
    r2 = r2_score(y_test, y_pred)
    
    results[name] = {
        "MAE": mae,
        "MSE": mse,
        "RMSE": rmse,
        "R2": r2
    }

# Compare Model Performance

In [30]:
import pandas as pd

performance_df = pd.DataFrame(results).T
performance_df.sort_values(by="RMSE")  # lower RMSE = better


,MAE,MSE,RMSE,R2
Random Forest,1.831507,14.091876,3.753915,0.997857
XGBoost,3.374459,22.854655,4.780654,0.996524
Decision Tree,2.332866,29.166690,5.400619,0.995564
KNN,5.303026,68.164110,8.256156,0.989633
Linear Regression,9.558810,166.188119,12.891397,0.974724


# Analysis

Lowest errors (MAE, RMSE):
Random Forest has the smallest MAE (1.83) and RMSE (3.75).
Linear Regression is now clearly the worst (MAE 9.56, RMSE 12.89).
Highest R²:
Random Forest → 0.997857 (best)
Linear Regression → 0.974724 (lowest)
Other models:
Decision Tree slightly worse than Random Forest.
XGBoost performs decently but not as good as Random Forest.
KNN is the weakest here.

# Based on metrics, Random Forest has the lowest errors and highest R²

# Model save

In [31]:
import joblib
from sklearn.ensemble import RandomForestRegressor

# Train the best model
best_model = RandomForestRegressor(n_estimators=100, random_state=42)
best_model.fit(X_train, y_train)

# Save as .pkl
joblib.dump(best_model, "best_regression_model.pkl")
print("Random Forest saved as 'best_regression_model.pkl'")


Random Forest saved as 'best_regression_model.pkl'
